lib/mix/tasks/dialyze.ex

```diff
@@ -82,6 +82,9 @@ defmodule Mix.Tasks.Dialyze do
     else
       warnings ->
         print_warnings(warnings)
+        if (match?([_|_], warnings)) do
+          Mix.raise "Dialyzer reported #{length(warnings)} warnings"
+        end
```

Comment:
This returns nil from `run/1`, spec says `:ok`. I think it would be clearer to match on the else clause.